### PR TITLE
drop julia 0.4 + fix errors and deprecations in 0.6

### PR DESCRIPTION
**Pull request by CarloLucibello**
I'm leaving inline comments on a few changes I'm not sure about

### CODE REVIEW

File: test/glib.jl [hunk 1]
```diff
@@ -22,13 +22,14 @@ repr = Base.print_to_string(wrap) #should display properties
...
-@test Test2.GtkWidgetTest2 != Gtk.GtkWidgetLeaf
-@test Test2.GtkWidget == Gtk.GtkWidget
-nothing
+# TODO
+# module Test2   ### fails inside @testset
+#     import Gtk
+#     using Gtk.GLib
+#     const suffix = :Test2
+#     @Gtype GtkWidget Gtk.libgtk gtk_widget
+# end
```
Comment (CarloLucibello) [inline test/glib.jl]: this is not passing on julia 0.6. What is this for and how can I fix it?

File: src/text.jl [hunk 2]
```diff
@@ -77,7 +77,7 @@ function GtkTextIter(text::GtkTextBuffer, mark::GtkTextMark)
 end
 show(io::IO, iter::GtkTextIter) = print("GtkTextIter(...)")
 
-immutable GtkTextRange <: Range
+immutable GtkTextRange <: Range{Char}
```
Comment (CarloLucibello) [inline src/text.jl]: is this right?

File: src/GLib/GLib.jl [hunk 3]
```diff
@@ -62,21 +54,15 @@ importall .CompatGLib
...
-if VERSION >= v"0.5.0-dev+4612"
-    bytestring(s::Ptr{UInt8}) = unsafe_string(s)
-elseif VERSION >= v"0.5.0-dev+4200"
-    bytestring(s::Ptr{UInt8}) = String(s)
-else
-    bytestring(s::Ptr{UInt8}) = Base.bytestring(s)
+function bytestring(s::Ptr{UInt8}, own::Bool)
+    # unsafe_wrap(String, s, ccall(:strlen, Csize_t, (Ptr{UInt8},), s), own)
+    @assert own==false #println("own=$own")
+    return bytestring(s)
```
Comment (CarloLucibello) [inline src/GLib/GLib.jl]: 3-arguments unsafe_wrap gives deprecation in favor of unsafe_string, which has no 3 arguments method. Is it ok to just let handle this to unsafe_string(s)?

**Comment by tknopp** [comment]
great that you work on this. @vtjnash @timholy I am not using 0.6 and thus cannot really review this.

**Comment by lobingera** [comment]
I can can pull and test this later today (~4h).
Just one thing: isn't it too early to drop julia 0.4? 0.6 isn't finished and 0.5 doesn't have a bugfix release, so technically 0.4 is the 'stable' version...

**Comment by tknopp** [comment]
this is up to discussion.  I disagree that 0.5 is not stable. In general it would be good to preserve 0.4 compatibility but the question is how much effort this is (and how maintainable the compatibility things are).

A practical solution is to ask: Is somebody using Gtk with 0.4? I am not.

**Comment by lobingera** [comment]
I do.

**Comment by tknopp** [comment]
Then dropping 0.4 support is not ok

**Comment by lobingera** [comment]
I can move forward to 0.5, that's not an issue; i don't think i have other only-0.4 packages on dependency list.

**Comment by CarloLucibello** [comment]
wouldn't be fine for 0.4 users to just use the latest 0.4-compatible tagged version?

File: src/GLib/GLib.jl (hunk 3 above)
Comment (timholy) [inline src/GLib/GLib.jl]: Sure, just use `unsafe_string`. It's the same on 0.5 and 0.6.

File: src/GLib/GLib.jl [hunk 4]
```diff
@@ -62,21 +54,15 @@ importall .CompatGLib
...
 end
+bytestring(s::Ptr{UInt8}) = unsafe_string(s)
```
Comment (timholy) [inline src/GLib/GLib.jl]: I'd get rid of this.

File: src/text.jl (hunk 2 above)
Comment (timholy) [inline src/text.jl]: Exactly. As you surely noted, `getindex(::GtkTextRange, ::Int)` returns a `Char`, so `Char` it should be.

File: test/glib.jl (hunk 1 above)
Comment (timholy) [inline test/glib.jl]: I don't understand either. @bfredl or @vtjnash? (As deduced by `git blame`.)

**Comment by lobingera** [comment]
just for the record, i branched a master and pulled  git pull https://github.com/JuliaGraphics/Gtk.jl refs/pull/287/head
then:
```
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-dev.2970 (2017-02-25 12:47 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit c0a324d* (5 days old master)
|__/                   |  x86_64-linux-gnu

julia> using Gtk
INFO: Precompiling module Gtk.
julia: /home/lobi/julia06/src/codegen.cpp:505: jl_cgval_t::jl_cgval_t(const jl_cgval_t&, jl_value_t*, llvm::Value*): Assertion `isboxed || v.typ == typ || tindex' failed.

signal (6): Aborted
while loading /home/lobi/.julia/v0.6/Gtk/src/GLib/gtype.jl, in expression starting on line 45
raise at /build/eglibc-oGUzwX/eglibc-2.19/signal/../nptl/sysdeps/unix/sysv/linux/raise.c:56
abort at /build/eglibc-oGUzwX/eglibc-2.19/stdlib/abort.c:89
__assert_fail_base at /build/eglibc-oGUzwX/eglibc-2.19/assert/assert.c:92
__assert_fail at /build/eglibc-oGUzwX/eglibc-2.19/assert/assert.c:101
jl_cgval_t at /home/lobi/julia06/src/codegen.cpp:505 [inlined]
update_julia_type at /home/lobi/julia06/src/codegen.cpp:763
emit_expr at /home/lobi/julia06/src/codegen.cpp:4020
emit_assignment at /home/lobi/julia06/src/codegen.cpp:3632 [inlined]
emit_expr at /home/lobi/julia06/src/codegen.cpp:4030
emit_stmtpos at /home/lobi/julia06/src/codegen.cpp:3935 [inlined]
emit_function at /home/lobi/julia06/src/codegen.cpp:6096
jl_compile_linfo at /home/lobi/julia06/src/codegen.cpp:1187
jl_compile_for_dispatch at /home/lobi/julia06/src/gf.c:1612
jl_call_method_internal at /home/lobi/julia06/src/julia_internal.h:224 [inlined]
jl_apply_generic at /home/lobi/julia06/src/gf.c:2212
copy! at ./abstractarray.jl:559
...
```

it goes on for a while...

**Comment by CarloLucibello** [comment]
weird, this branch is working on my linux box and on travis....

**Comment by lobingera** [comment]
0.6.0-dev.2970?

**Comment by CarloLucibello** [comment]
```
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-pre.alpha.5 (2017-02-28 18:33 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit b1242dff00 (1 day old master)
|__/                   |  x86_64-pc-linux-gnu
```

**Comment by lobingera** [comment]
I updated v0.6 and get now (similar):
```
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-pre.alpha.20 (2017-03-02 17:16 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 2bc66e1* (0 days old master)
|__/                   |  x86_64-linux-gnu

julia> using Gtk
INFO: Precompiling module Gtk.
julia: /home/lobi/julia06/src/codegen.cpp:505: jl_cgval_t::jl_cgval_t(const jl_cgval_t&, jl_value_t*, llvm::Value*): Assertion `isboxed || v.typ == typ || tindex' failed.

signal (6): Aborted
while loading /home/lobi/.julia/v0.6/Gtk/src/GLib/gtype.jl, in expression starting on line 45
raise at /build/eglibc-oGUzwX/eglibc-2.19/signal/../nptl/sysdeps/unix/sysv/linux/raise.c:56
abort at /build/eglibc-oGUzwX/eglibc-2.19/stdlib/abort.c:89
__assert_fail_base at /build/eglibc-oGUzwX/eglibc-2.19/assert/assert.c:92
__assert_fail at /build/eglibc-oGUzwX/eglibc-2.19/assert/assert.c:101
jl_cgval_t at /home/lobi/julia06/src/codegen.cpp:505 [inlined]
update_julia_type at /home/lobi/julia06/src/codegen.cpp:763
emit_expr at /home/lobi/julia06/src/codegen.cpp:4020
emit_assignment at /home/lobi/julia06/src/codegen.cpp:3632 [inlined]
emit_expr at /home/lobi/julia06/src/codegen.cpp:4030
emit_stmtpos at /home/lobi/julia06/src/codegen.cpp:3935 [inlined]
emit_function at /home/lobi/julia06/src/codegen.cpp:6096
jl_compile_linfo at /home/lobi/julia06/src/codegen.cpp:1187
jl_compile_for_dispatch at /home/lobi/julia06/src/gf.c:1612
jl_call_method_internal at /home/lobi/julia06/src/julia_internal.h:224 [inlined]
jl_apply_generic at /home/lobi/julia06/src/gf.c:2212
copy! at ./abstractarray.jl:559
collect at ./array.jl:347
unknown function (ip: 0x7f3b7e48dfb4)
jl_call_method_internal at /home/lobi/julia06/src/julia_internal.h:248 [inlined]
jl_apply_generic at /home/lobi/julia06/src/gf.c:2212
do_call at /home/lobi/julia06/src/interpreter.c:75
eval at /home/lobi/julia06/src/interpreter.c:230
eval_body at /home/lobi/julia06/src/interpreter.c:534
jl_interpret_toplevel_thunk at /home/lobi/julia06/src/interpreter.c:679
jl_toplevel_eval_flex at /home/lobi/julia06/src/toplevel.c:592
jl_parse_eval_all at /home/lobi/julia06/src/ast.c:873
jl_load at /home/lobi/julia06/src/toplevel.c:616
include_from_node1 at ./loading.jl:539
unknown function (ip: 0x7f3b7e44dff0)
jl_call_method_internal at /home/lobi/julia06/src/julia_internal.h:248 [inlined]
jl_apply_generic at /home/lobi/julia06/src/gf.c:2212
include at ./sysimg.jl:14
unknown function (ip: 0x7f3d8f7a893b)
jl_call_method_internal at /home/lobi/julia06/src/julia_internal.h:248 [inlined]
jl_apply_generic at /home/lobi/julia06/src/gf.c:2212
do_call at /home/lobi/julia06/src/interpreter.c:75
eval at /home/lobi/julia06/src/interpreter.c:230
jl_interpret_toplevel_expr at /home/lobi/julia06/src/interpreter.c:34
jl_toplevel_eval_flex at /home/lobi/julia06/src/toplevel.c:577
jl_eval_module_expr at /home/lobi/julia06/src/toplevel.c:203
jl_toplevel_eval_flex at /home/lobi/julia06/src/toplevel.c:480
jl_parse_eval_all at /home/lobi/julia06/src/ast.c:873
jl_load at /home/lobi/julia06/src/toplevel.c:616
include_from_node1 at ./loading.jl:539
unknown function (ip: 0x7f3d8f8f8c8b)
jl_call_method_internal at /home/lobi/julia06/src/julia_internal.h:248 [inlined]
jl_apply_generic at /home/lobi/julia06/src/gf.c:2212
include at ./sysimg.jl:14
unknown function (ip: 0x7f3d8f7a893b)
jl_call_method_internal at /home/lobi/julia06/src/julia_internal.h:248 [inlined]
jl_apply_generic at /home/lobi/julia06/src/gf.c:2212
do_call at /home/lobi/julia06/src/interpreter.c:75
eval at /home/lobi/julia06/src/interpreter.c:230
jl_interpret_toplevel_expr at /home/lobi/julia06/src/interpreter.c:34
jl_toplevel_eval_flex at /home/lobi/julia06/src/toplevel.c:577
jl_eval_module_expr at /home/lobi/julia06/src/toplevel.c:203
jl_toplevel_eval_flex at /home/lobi/julia06/src/toplevel.c:480
jl_parse_eval_all at /home/lobi/julia06/src/ast.c:873
jl_load at /home/lobi/julia06/src/toplevel.c:616
include_from_node1 at ./loading.jl:539
unknown function (ip: 0x7f3d8f8f8c8b)
jl_call_method_internal at /home/lobi/julia06/src/julia_internal.h:248 [inlined]
jl_apply_generic at /home/lobi/julia06/src/gf.c:2212
include at ./sysimg.jl:14
unknown function (ip: 0x7f3d8f7a893b)
jl_call_method_internal at /home/lobi/julia06/src/julia_internal.h:248 [inlined]
jl_apply_generic at /home/lobi/julia06/src/gf.c:2212
do_call at /home/lobi/julia06/src/interpreter.c:75
eval at /home/lobi/julia06/src/interpreter.c:230
jl_interpret_toplevel_expr at /home/lobi/julia06/src/interpreter.c:34
jl_toplevel_eval_flex at /home/lobi/julia06/src/toplevel.c:577 [inlined]
jl_toplevel_eval at /home/lobi/julia06/src/toplevel.c:600
jl_toplevel_eval_in at /home/lobi/julia06/src/builtins.c:484
anonymous at ./<missing>:2
jl_call_method_internal at /home/lobi/julia06/src/julia_internal.h:248 [inlined]
jl_toplevel_eval_flex at /home/lobi/julia06/src/toplevel.c:589 [inlined]
jl_toplevel_eval at /home/lobi/julia06/src/toplevel.c:600
jl_toplevel_eval_in at /home/lobi/julia06/src/builtins.c:484
eval at ./boot.jl:235
unknown function (ip: 0x7f3d8f8cce0f)
jl_call_method_internal at /home/lobi/julia06/src/julia_internal.h:248 [inlined]
jl_apply_generic at /home/lobi/julia06/src/gf.c:2212
process_options at ./client.jl:286
_start at ./client.jl:371
unknown function (ip: 0x7f3d8f9269c8)
jl_call_method_internal at /home/lobi/julia06/src/julia_internal.h:248 [inlined]
jl_apply_generic at /home/lobi/julia06/src/gf.c:2212
jl_apply at /home/lobi/julia06/ui/../src/julia.h:1410 [inlined]
true_main at /home/lobi/julia06/ui/repl.c:127
main at /home/lobi/julia06/ui/repl.c:264
__libc_start_main at /build/eglibc-oGUzwX/eglibc-2.19/csu/libc-start.c:287
unknown function (ip: 0x4015fa)
Allocations: 3547064 (Pool: 3545477; Big: 1587); GC: 4
ERROR: Failed to precompile Gtk to /home/lobi/.julia/lib/v0.6/Gtk.ji.
Stacktrace:
 [1] compilecache(::String) at ./loading.jl:673
 [2] require(::Symbol) at ./loading.jl:460

julia> 
```

File: .travis.yml [hunk 5]
```diff
@@ -4,7 +4,6 @@ os:
   - linux
   - osx
 julia:
-  - 0.4
```
Comment (tkelman) [inline .travis.yml]: you need to change the minimum julia version in REQUIRE

File: REQUIRE [hunk 6]
```diff
@@ -4,5 +4,5 @@ Graphics 0.1
 BinDeps
 @windows WinRPM
 @osx Homebrew
-Compat 0.9.5
+Compat 0.19
```
Comment (tkelman) [inline REQUIRE]: 0.18, since you're not using `IndexStyle`

**Comment by timholy** [comment]
AFAICT, this is ready to go except for that one commented-out tests. I'm inclined to merge anyway. Any objections?

File: test/glib.jl (hunk 1 above)
Comment (vtjnash) [inline test/glib.jl]: It's testing the macro hygiene of `@Gtype`, which is apparently broken :P

yeah, not worth holding up fixing everything else for this though.

**Comment by CarloLucibello** [comment]
probably there is some crust that could be removed, e.g. `bytestring`, but this as a far as I dare go since I'm not familiar with the package. 👍

File: src/GLib/GLib.jl [hunk 7]
```diff
@@ -62,21 +54,11 @@ importall .CompatGLib
 using .CompatGLib.WORD_SIZE
 
 # local function, handles Symbol and makes UTF8-strings easier
-typealias AbstractStringLike Union{AbstractString, Symbol}
+@compat const  AbstractStringLike = Union{AbstractString, Symbol}
```
Comment (tkelman) [inline src/GLib/GLib.jl]: compat not needed for this

File: src/GLib/gtype.jl [hunk 8]
```diff
@@ -1,12 +1,12 @@
...
+@compat abstract type GInterface <: GObject end
+@compat abstract type GBoxed  end
 type GBoxedUnkown <: GBoxed
     handle::Ptr{GBoxed}
 end
 
-typealias GEnum Int32
-typealias GType Csize_t
+@compat const  GEnum = Int32
+@compat const  GType = Csize_t
```
Comment (tkelman) [inline src/GLib/gtype.jl]: compat not needed

**Pull request merged**